mlos_core/mlos_core/data_classes.py

```diff
@@ -0,0 +1,350 @@
+        ----------
+        index : pd.Index
+            The indices to filter by.
+
+        Returns
+        -------
+        Observation
+            The filtered observation.
+        """
+        return Observations(
+            config=self._config.loc[index].copy(),
+            score=self._score.loc[index].copy(),
+            context=None if self._context is None else self._context.loc[index].copy(),
+            metadata=None if self._metadata is None else self._metadata.loc[index].copy(),
+        )
+
+    def append(self, other: Observation) -> None:
+        """
+        Appends the given observation to this observation.
+
+        Parameters
+        ----------
+        other : Observation
+            The observation to append.
+        """
+        config = other._config.to_frame().T
+        score = other._score.to_frame().T
+        context = None if other._context is None else other._context.to_frame().T
+        metadata = None if other._metadata is None else other._metadata.to_frame().T
+        if len(self._config.index) == 0:
+            self._config = config
+            self._score = score
+            self._context = context
+            self._metadata = metadata
+            assert set(self._config.index) == set(
+                self._score.index
+            ), "config and score must have the same index"
+            return
+
+        self._config = pd.concat([self._config, config]).reset_index(drop=True)
+        self._score = pd.concat([self._score, score]).reset_index(drop=True)
+        assert set(self._config.index) == set(
+            self._score.index
+        ), "config and score must have the same index"
+
+        if self._context is not None:
+            assert context is not None, (
+                "context of appending observation must not be null"
+                + " if context of prior observation is not null"
+            )
+            self._context = pd.concat([self._context, context]).reset_index(drop=True)
+            assert self._config.index.equals(
+                self._context.index
+            ), "config and context must have the same index"
+        else:
+            assert context is None, (
+                "context of appending observation must be null"
+                + " if context of prior observation is null"
+            )
+        if self._metadata is not None:
+            assert metadata is not None, (
+                "context of appending observation must not be null"
+                + " if metadata of prior observation is not null"
+            )
+            self._metadata = pd.concat([self._metadata, metadata]).reset_index(drop=True)
+            assert self._config.index.equals(
+                self._metadata.index
+            ), "config and metadata must have the same index"
+        else:
+            assert metadata is None, (
+                "context of appending observation must be null"
+                " if metadata of prior observation is null"
```

Comment:
Above you use an explicit concat and here you use an implicit one.
Maybe just pick one style and stick to it?

Reply:
Also, preference for trailing whitespace on the previous line.